### ISSUE TITLE
Need to show no stance differently from disagreement in matching

### ISSUE THREAD

**Issue by crieggalder** [andreisarebi/civic-app#83]
3 issues with matching:

1. After playing with candidates for a bit, I really think we should represent the case where the user has a stance but the candidate doesn't differently than disagreement. It's counter-intuitive and probably misleading for the user to read "Disagree on [x] issues", click the item, and see "[A] has not expressed a clear stance on this issue." I suggest we show a third symbol - a gray question mark - to represent areas where the candidate has not expressed a stance.

2. It seems that candidates without stances on many issues are showing high match percentages because the gap is "only" from 0 (no stance) to the user's position (+2 if very progressive like me), rather than the max possible gap of -2 (very conservative) to +2 (very progressive). As a result, a candidate like John Fitzgerald (fringe conspiracy theorist) shows as a 44% match despite having no stance on 7 of 9 issues and disagreeing with me on the other two. I propose we exclude issues where the candidate has no position from the match calculation.

3. Let's show match % without the decimals -- I think they distract the user. :)

![86708593-8467-48d1-80fd-f228109485b8](https://user-images.githubusercontent.com/36721640/46699245-1329ad00-cbce-11e8-8da5-daf1f020980a.png)

**Comment by crieggalder** [comment]
@cgalbiati thought more about it and I think match % should be based on number of issues where the user matches the candidate, not so much around intensity of agreement. How about this?

1. Calculate agreement value: 1 point for each issue where user and candidate have same position. Agree vs. strongly agree doesn't matter, 9 points possible.
2. Calculate disagreement value: 1 point for each issue where user and candidate have opposite positions. Disagree vs. strongly disagree doesn't matter, 9 points possible.
3. ( [agreement value] - [disagreement value] ) / 9. Normalize so that 1 = 100% match, 0 = 50% match, -1 = 0% match.

Drawback is that this doesn't weight by intensity of agreement, e.g. if I care most about the environment and disagree with a candidate, but agree on something I care less about, both issues have equal weight. But I'm struggling to find a way to account for that without showing candidates with whom I weakly agree across the board as only 50% matches (50% of possible agreement points). IMO they should be 100% matches.

**Comment by crieggalder** [comment]
Nvm, above is still wrong because a candidate with no stances would show as 50% again. Gotta account for that. Will think on it more.

**Comment by crieggalder** [comment]
@cgalbiati, I think we solve the problem of candidates without positions by treating them as disagreement for the purposes of matching. Calculation:
1. Calculate agreement value: 1 point for each issue where user and candidate have same position. Agree vs. strongly agree doesn't matter, 9 points possible.
4. ( [agreement value] - [9 - agreement value] ) / 9. Normalize so that 1 = 100% match, 0 = 50% match, -1 = 0% match.

That fixes our issue with stance-less candidates. It doesn't weight by intensity - thoughts on how we could that?

**Comment by cgalbiati** [comment]
@crieggalder several thoughts:
1. We can easily round the viewed match %'s. (just assign it to someone)
2. We can easily add a new icon - it is actually already set up to take one in the data layer, we just didn't have the icon, so the view would need some minor adjustments.

3. Amount of agreement/disagreement: We actually already do not take into account intensity of agreement, since we don't show range for candidates, but I think we should keep the distinction for disagreement - I do think disagreeing on an issue where a user has strong feelings should make lower your match more than an issue you don't care that much about.  If we don't take into account the intensity of the disagreement at all, I don't think we should include it in the survey.    However, we could weight it differently, so that disagreeing less intensely is less different from disagreeing intensely.  
I might propose it should look more like agree on all = 100, no position = 50, strongly disagree=0, weakly disagree=30.  That would essentially mean that agreeing on 8 and strongly disagreeing on 2 (80% match) would be approximately the same as agreeing on 7 and weakly disagreeing on 3 (79% match).  We could adjust the difference if you think it should be weighted differently, but I do think they should be different.  We actually might want to show that they strongly disagree in the breakdown by changing the copy (I don't think we would need to change the icon)

4. We can discount issues where a candidate has not expressed a stance.  I would probably argue for still including them, in some way, since I should have a higher match with someone who has expressed more positions that I agree with, vs someone who may or may not agree with me on those positions.  We could weight those as less value though. I think that 8 agree + 2 strongly agree should maybe be the same as 5 agree + 5 unknown?

Overall, the system is set up to make it pretty easy to change the weights of everything.  I would actually propose making a couple of versions and testing them on some users, so that we can see what they think of the % when they see the breakdown, and see if they think it makes sense.  We could even leave the match% out and give people different combinations of agree, no position, disagree, strongly disagree, and have them rank the candidates on which they think they would be most likely to vote for.

**Comment by mangowango** [comment]
@crieggalder ,

Have been doing some research into optimizing the match algorithm! Can you explain why we only capture candidates on a -1,1 scale again? Is it too tricky to give them the same -2,2 as users? 

Also, have been researching two types of scoring / matching algs: jaccard and cosine similarity.  Both seem pretty ideal for our use case. I think we should benchmark our custom algs and weightings against these techniques that are pretty accepted in the DS community. If we have a technique that is better out of the box, I think we maybe just roll with that instead of reinventing the wheel. Also @cgalbiati 's third point above is really good, I'll try to work out some sample versions of weighting later this week....also-also, cool if I write some of this in python? lol 


Jaccard evaluates sets, flexible enough to account for both magnitude and direction, I think 

Cosine doesn't take into account magnitude, which seems like it makes sense given our mismatch between candidates scoring and user scoring

Also, for a little bit of added visibility, I'm gunna test against something like this that I mocked up. I made a few types of users and a few types of candidates that I think rep both edge and expected use cases.

[simple_matching_model.xlsx](https://github.com/civic-app/civic-app/files/2479821/simple_matching_model.xlsx)

LMK what y'all think!

**Comment by crieggalder** [comment]
Hey JoJo! Awesome work!

Yes, we only capture candidates from -1 to 1 because it's too difficult to
distinguish support from strong support.

Algos sound cool! How would they take into account lack of information for
some/all issues on candidate side?

Great stuff!

**Comment by mangowango** [comment]
algos for days! love me some code code names hahahah.  And when handling missing data, neither alg intuitively processes missing data, we've gotta make some decisions there.  

Some basics we can try:

- we can replace missing data with zeros
-  the median, mean, or mode

Next level:

- use a k nearest neighbor (KNN) classification or expected max (EM) alg to fill in the gaps. Would think KNN makes more intuitive sense, because it basically means hey, find a candidate that is similar and use their data for now and is much less expensive in compute power than EM, I believe. Also it seems to be a good jumping off point for missing data in DS projects I've seen.

Basically, all roads seem to lead back to "let's test cases for missing data". The great thing is that once we build one alg, swapping in different methods for missing data is easy. Def don't want to go overboard tho because I feel like these methods should be easy just to give us some benchmarks before we take a deep dive and construct our own logic.

**Comment by mangowango** [comment]
Developed the first baseline alg with Cosine, am gunna do Jaccard soon, and then build out Chandra's suggestions too!
[cosine_matching_alg.txt](https://github.com/civic-app/civic-app/files/2489427/cosine_matching_alg.txt)
